### PR TITLE
KEYCLOAK-12526: Add 'Always Display in Console' to admin console

### PR DESCRIPTION
**Pull request by ssilvert**

### CODE REVIEW

**Comment by ssilvert** [comment]
@douglaspalmer Please verify that the tooltip text is correct.

**Review by abstractj — CHANGES_REQUESTED** [review]
@ssilvert great stuff, I believe we will need to update https://github.com/keycloak/keycloak/blob/65f51b7b83271bc42de923b3625cac8a69e8e9a2/testsuite/integration-arquillian/tests/other/console/src/main/java/org/keycloak/testsuite/console/page/clients/settings/ClientSettingsForm.java to something like:

```
@FindBy(xpath = ".//div[@class='onoffswitch' and ./input[@id='alwaysDisplayInConsole']]")
private OnOffSwitch alwaysDisplayInConsole;
```

After that I believe we need to update the tests here: https://github.com/keycloak/keycloak/blob/24e60747b694ab4d03e8e1cbf8e4da764337ff48/testsuite/integration-arquillian/tests/other/console/src/test/java/org/keycloak/testsuite/console/clients/ClientSettingsTest.java

At the moment UI tests are not running on Travis, but as far as I know we have it enabled on Jenkins.

File: themes/src/main/resources/theme/base/admin/resources/partials/client-detail.html
Comment (douglaspalmer) [inline themes/src/main/resources/theme/base/admin/resources/partials/client-detail.html]: Should we make this invisible for bearer only clients and those with no base url. The users may get confused if they set alwaysDisplayInConsole and then the client isn't displayed.

**Review by douglaspalmer — CHANGES_REQUESTED** [review]

File: themes/src/main/resources/theme/base/admin/resources/partials/client-detail.html
Comment (ssilvert) [reply themes/src/main/resources/theme/base/admin/resources/partials/client-detail.html]: @douglaspalmer @abstractj Yes, bearer only clients need to be handled properly in the UI.  I think it will need a trigger so that it sets alwaysDisplayInConsole to false and disables the selection whenever bearer only is set.

I'm not sure about the "no base url" thing.  Are we unable to show in account console if there is no base url?

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/admin/ClientTest.java
Comment (ssilvert) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/admin/ClientTest.java]: That line is just there for completeness.  We will do the actual test in the console test suggested by Bruno.

**Comment by ssilvert** [comment]
> @ssilvert great stuff, I believe we will need to update https://github.com/keycloak/keycloak/blob/65f51b7b83271bc42de923b3625cac8a69e8e9a2/testsuite/integration-arquillian/tests/other/console/src/main/java/org/keycloak/testsuite/console/page/clients/settings/ClientSettingsForm.java to something like:
> 
> ```
> @FindBy(xpath = ".//div[@class='onoffswitch' and ./input[@id='alwaysDisplayInConsole']]")
> private OnOffSwitch alwaysDisplayInConsole;
> ```
> 
> After that I believe we need to update the tests here: https://github.com/keycloak/keycloak/blob/24e60747b694ab4d03e8e1cbf8e4da764337ff48/testsuite/integration-arquillian/tests/other/console/src/test/java/org/keycloak/testsuite/console/clients/ClientSettingsTest.java
> 
> At the moment UI tests are not running on Travis, but as far as I know we have it enabled on Jenkins.

@abstractj @vmuzikar I can't get ClientSettingsTest#crudOIDCPublic to run.  Can either of you verify if it works?

I get:
ERROR [com.gargoylesoftware.htmlunit.javascript.StrictErrorReporter] runtimeError: message=[An invalid or illegal selector was specified (selector: '*,:x' error: Invalid selector: :x).] sourceName=[https://localhost:8543/auth/resources/remck/admin/keycloak/node_modules/jquery/dist/jquery.min.js] line=[2] lineSource=[null] lineOffset=[0]

Then htmlunit throws a NullPointerException.

**Comment by vmuzikar** [comment]
@ssilvert I can confirm `ClientSettingsTest` passes with your changes. How exactly are you executing the test? Htmlunit is not supposed to even run with the Admin Console tests (it should be blocked on POM level), I wonder how did you manage to use htmlunit here. :) You need to use a [real life browser](https://github.com/keycloak/keycloak/blob/master/testsuite/integration-arquillian/HOW-TO-RUN.md#different-browsers).

We'll need to extend the test to at least check if the new switch does what it's supposed to (if it really changes the client config, nothing more). But preferably we should also check if the switch is hidden when the `account2` profile is not active.

**Review by abstractj — APPROVED** [review]
I managed to run all the tests and nothing was broken locally, so I'm approving.

**Review by abstractj — CHANGES_REQUESTED** [review]
Based on our discussion today, it requires more changes and I overlooked this PR.

Sorry about that.

**Comment by ssilvert** [comment]
@abstractj @douglaspalmer I made the changes you requested.  Ready for another round of review.

File: testsuite/integration-arquillian/tests/other/console/pom.xml
Comment (vmuzikar) [inline testsuite/integration-arquillian/tests/other/console/pom.xml]: I believe we really shouldn't hardcode the profile here. This way we'd run all Admin Console tests with a preview profile – always, even when e.g. testing product.

I've been a bit playing with this and I think we should use `@EnableFeature` annotation for a separate test just for the new switch. You can use my changes if you want: https://github.com/ssilvert/keycloak/pull/4

**Review by vmuzikar — CHANGES_REQUESTED** [review]
@ssilvert Thanks for the update.

File: themes/src/main/resources/theme/base/admin/resources/partials/client-detail.html
Comment (douglaspalmer) [reply themes/src/main/resources/theme/base/admin/resources/partials/client-detail.html]: Wouldn't it be better to fix the code and remove the bearer-only flag when the client is switched to SAML?

Comment (vmuzikar) [reply themes/src/main/resources/theme/base/admin/resources/partials/client-detail.html]: @douglaspalmer I don't see it as a bug (that the bearer-only flag is preserved even for SAML). This is a deeper issue that we 1) allow to change protocol of an existing client and 2) we use the same client representation for OIDC and SAML (and docker clients). I mean there are plenty of OIDC or SAML specific attributes that are stored even when not relevant for given protocol.

File: testsuite/integration-arquillian/tests/other/console/pom.xml
Comment (ssilvert) [reply testsuite/integration-arquillian/tests/other/console/pom.xml]: I agree.  Didn't know about the @EnableFeature annotation.  That's much better.

**Comment by ssilvert** [comment]
Changed from @vmuzikar are now merged into this PR.

@abstractj @douglaspalmer Ready for (hopefully) final review.

File: themes/src/main/resources/theme/base/admin/resources/partials/client-detail.html
Comment (vmuzikar) [reply themes/src/main/resources/theme/base/admin/resources/partials/client-detail.html]: @ssilvert The tests currently expect this to be fixed. Should we fix this or should we update the test? :)

**Review by douglaspalmer — COMMENTED** [review]
I think this is good to go once the Bearer-Only/SAML issue is fixed.

**Comment by ssilvert** [comment]
> I think this is good to go once the Bearer-Only/SAML issue is fixed.

I agree with @vmuzikar on this one.  I don't think we need a change because the same issue exists for lots of attributes.  Really, it's a fundamental problem with the data implementation.

Maybe @abstractj can be the tie breaker, but IMO we should merge.

File: themes/src/main/resources/theme/base/admin/resources/partials/client-detail.html
Comment (ssilvert) [reply themes/src/main/resources/theme/base/admin/resources/partials/client-detail.html]: @vmuzikar I'll fix the code.

Comment (ssilvert) [reply themes/src/main/resources/theme/base/admin/resources/partials/client-detail.html]: @vmuzikar @douglaspalmer Fixed now.

**Review by douglaspalmer — APPROVED** [review]
LGTM now

**Review by vmuzikar — APPROVED** [review]
LGTM, thanks @ssilvert!

**Review by abstractj — APPROVED** [review]

**Pull request merged**
